### ISSUE TITLE
WCF server receives the message, but client registers error

### ISSUE THREAD

**Issue by Peperud** [microsoft/dotnet#1115]
I wonder how reliable is the WCF protocol with ```netTcpBinding```?

We have a producer - consumer scenario where the producer is a WCF client and the consumer is one of several WCF servers. If the producer fails to send a job to a consumer, it tries again with another consumer and so forth.

Looks like sometimes the producer sends the job, a consumer gets it fine, however the producer registers a network error (guessing a temporary network glitch causing it to never get the ack from the server that the call succeeded) and then attempts to send the same job to another consumer. 

Does that sound right/possible?

I'm not sure whether WCF (by itself) is reliable enough for the described (perhaps somewhat naive) protocol?

**Comment by vivmishra** [comment]
@StephenBonikowsky

**Comment by StephenBonikowsky** [comment]
@Peperud depending on how WCF is configured on your client and server it can be very reliable.
For instance, WCF on full framework supports the [WS-ReliableMessaging specification](http://docs.oasis-open.org/ws-rx/wsrm/200608/wsrm-1.1-spec-cd-04.html).

The `netTcpBinding` supports this via the [NetTcpBinding.ReliableSession Property](https://docs.microsoft.com/en-us/dotnet/api/system.servicemodel.nettcpbinding.reliablesession?view=netframework-4.8#System_ServiceModel_NetTcpBinding_ReliableSession).

What you described seems like expected behavior where reliable messaging is not setup and therefore the communications are susceptible to network failures.

@mconnew For the described scenario would you recommend some other approach to solving their problem?

**Comment by Peperud** [comment]
@StephenBonikowsky 
Thanks. Yes, I noticed that reliable messaging was not turned on. It will be in the next release.

Reading the overview though, it looks to me that reliable messaging is more about the client retrying (certain number of times) if it does not receive ac ack, and the server (eventually) ignoring duplicates. Not about "_reliably failing_" on both sides. Which seems is what I need.

I hope I'm wrong, but one could speculate that it would still be possible for the client to send a message, the server to receive it and start working on it and the client (due to network etc.) not to receive the ack from the server (even after several retries), and then fail and send the same job to another server...

Or is the failure on both sides guaranteed?

**Comment by StephenBonikowsky** [comment]
It sounds like you need to make your services [Idempotent](https://en.wikipedia.org/wiki/Idempotence). That is a design solution at the service or application layer not the message or transport layer.

So yes, enabling ReliableMessaging won't help your specific scenario. There are many examples out there on how to setup operations to be Idempotent.

A good starting point for initial investigation/understanding of the concept would be searching for Idempotent on https://stackexchange.com/

**Comment by Peperud** [comment]
@StephenBonikowsky 
Thanks for confirming that WCF cannot be trusted to fail reliably.

**Comment by mconnew** [comment]
@Peperud, can you please explain what you mean by fail reliably?

**Comment by Peperud** [comment]
@mconnew  What I mean by _fail reliably_ is to never have a situation where the client registers a fault, but the server doesn't.

**Comment by mconnew** [comment]
@Peperud, that's not possible to guarantee with any stack without building more functionality out in the application layer code. Take the situation where there's a network problem after the service side code completes the operation but before it has sent the response to the client. At that point your service code has completed but the response is unable to be sent to the client which the client will become aware of, but the service implementation is not.  

With the NetTcp transport WCF does have some capability for you to add a handler to know when the connection has failed after your operation has completed. In your service code you can add an event handler to the channel which gets called when the channel has faulted. Off the top of my head I can't think of a way to let you know whether the channel faulting happened before or after a particular response was sent to the client though.  

There are other techniques you can use to have the client effectively inform the service when it failed. If you use distributed transactions, you start a transaction before the client makes the call, make the call, then if the client fails to get a response, you rollback the transaction. On the service side, you use the rollback as notification that the client failed getting the response. If your app is only making state changes which participate in transactions, e.g. you are changing data in a SQL database, there isn't much you need to do other than some service configuration to make the transaction flow and configure your sql queries to be part of the transaction. If you have a more complicated scenario, you will need to write a transaction participant to get notified and take whatever action you want.  

Basically with distributed computing, you have two options. Either make it idempotent like Stephen mentioned, or do full distributed transactions. This isn't a limitation of WCF, this is the nature of distributed programming.

**Comment by Peperud** [comment]
@mconnew, Yes, I understand that it is like that by its nature and this is not limited to WCF. I was hoping that WCF might have something in its sleeve to make it easier. I guess the distributed transaction support counts. Thanks.
